File: src/com/eagle/programmar/Rust/Rust_Function.java
// Copyright Eagle Legacy Modernization, LLC, 2010-date
// Original author: Steven A. O'Hara, Jul 2, 2022

package com.eagle.programmar.Rust;

import java.util.ArrayList;
import java.util.Collection;

import com.eagle.generate.EagleGenerator;
import com.eagle.generate.TypeEnum;
import com.eagle.interpret.EagleInterpreter;
import com.eagle.interpret.EagleRunnable;
import com.eagle.metrics.ArgumentsMetrics;
import com.eagle.metrics.CallMetrics;
import com.eagle.programmar.Rust.Rust_Type.Rust_TypePrimitive;
import com.eagle.programmar.Rust.Statements.Rust_Block_Statement;
import com.eagle.programmar.Rust.Symbols.Rust_Function_Definition;
import com.eagle.programmar.Rust.Symbols.Rust_Variable_Definition;
import com.eagle.programmar.Rust.Terminals.Rust_Comment;
import com.eagle.programmar.Rust.Terminals.Rust_Keyword;
import com.eagle.programmar.Rust.Terminals.Rust_Punctuation;
import com.eagle.scope.EagleScope;
import com.eagle.scope.EagleScope.EagleScopeInterface;
import com.eagle.tokens.AbstractFunction;
import com.eagle.tokens.SeparatedList;
import com.eagle.tokens.TokenList;
import com.eagle.tokens.TokenSequence;
import com.eagle.tokens.interfaces.AbstractExpression;
import com.eagle.tokens.interfaces.AbstractStatement;
import com.eagle.tokens.interfaces.AbstractType;
import com.eagle.tokens.interfaces.AbstractVariable;
import com.eagle.tokens.punctuation.PunctuationColon;
import com.eagle.tokens.punctuation.PunctuationComma;
import com.eagle.tokens.punctuation.PunctuationLeftBrace;
import com.eagle.tokens.punctuation.PunctuationLeftParen;
import com.eagle.tokens.punctuation.PunctuationRightBrace;
import com.eagle.tokens.punctuation.PunctuationRightParen;
import com.eagle.transform.EagleTransformableFunction;
import com.eagle.transform.EagleTransformer;

public class Rust_Function extends TokenSequence
		implements EagleRunnable, AbstractFunction, EagleScopeInterface, EagleTransformableFunction
{
	public @S(10) @OPT @BLANKLINE Rust_Keyword PUB = new Rust_Keyword("pub");
	public @S(20) @DOC("items/functions.html") Rust_Keyword FN = new Rust_Keyword("fn");
	public @S(30) Rust_Function_Definition id;
	public @S(40) @NOSPACE PunctuationLeftParen leftParen;
	public @S(50) @OPT @NOSPACE SeparatedList<Rust_Parameter, PunctuationComma> funcParamDefs;
	public @S(60) @NOSPACE PunctuationRightParen rightParen;
	public @S(70) @OPT Rust_FunctionReturns returns;
	public @S(80) Rust_Block_Statement block;

	public static class Rust_FunctionReturns extends TokenSequence
	{
		public @S(10) Rust_Punctuation arrow = new Rust_Punctuation("->");
		public @S(20) Rust_Type returnType;
	}

	public static class Rust_Parameter extends TokenSequence
	{
		public @S(10) Rust_Variable_Definition var;
		public @S(20) PunctuationColon colon;
		public @S(30) Rust_Type type;
	}

	private @SKIP EagleScope _scope = new EagleScope(this, Rust_Syntax.IS_CASE_SENSITIVE);
	public @SKIP CallMetrics _callMetrics = null;
	public @SKIP ArgumentsMetrics _argumentsMetrics = null;

	@Override
	public EagleScope getScope()
	{
		return _scope;
	}

	@Override
	public void interpret(EagleInterpreter interpreter)
	{
		if (_callMetrics == null)
		{
			_callMetrics = new CallMetrics(interpreter._metrics, id.getValue(), id);
		}
		if (_argumentsMetrics == null)
		{
			_argumentsMetrics = new ArgumentsMetrics(interpreter._metrics, id.getValue(), id);
		}

		// Don't do anything here.
		// We searched for all the functions in a preliminary pass
		// And we only evaluate when it is called

		// Except the function called 'main'
		if (id.getValue().equals("main"))
		{
			interpreter.callingFunction("main", this);
			interpreter.tryToInterpret(block);
			interpreter.completedFunction("main", this);
		}
	}

	@Override
	public void transformFunction(EagleTransformer transformer,
			EagleGenerator<AbstractStatement, AbstractExpression, AbstractVariable, AbstractType> generator)
	{
		AbstractType newReturnType = null;
		if (returns != null && returns.isPresent())
		{
			if (returns.returnType.getWhich() instanceof Rust_TypePrimitive)
			{
				Rust_TypePrimitive primitive = (Rust_TypePrimitive) returns.returnType.getWhich();
				newReturnType = Rust_Type.findType(generator, primitive.PRIMITIVE.toString());
			}
		}

		if (newReturnType == null)
		{
			TypeEnum metricRetType = transformer.findReturnMetric(id);
			newReturnType = generator.transformType(metricRetType, null, id);
		}

		String fnName = id.getValue();
		boolean isMain = false;
		if (fnName.equals("main"))
		{
			fnName = generator.mainName(); // Change from 'main' to 'Main' for C#
			isMain = true;
		}

		generator.addMethod(newReturnType, fnName, this);
		generator.addMethodName(fnName);
		if (VERBOSE)
		{
			System.out.println("** Found Rust function " + fnName);
		}

		if (isMain)
		{
			// Have to wait until addMethod is called
			generator.addMainArgs(); // For java and C# but not for Python or Rust
		}

		// Search metrics for arg types -- might not be any
		ArrayList<TypeEnum> argTypes = transformer.findArgumentsMetric(id);

		if (funcParamDefs != null && funcParamDefs.isPresent())
		{
			for (int i = 0; i < funcParamDefs.getPrimaryCount(); i++)
			{
				Rust_Parameter param = funcParamDefs.getPrimaryElement(i);
				AbstractType paramType = null;

				if (argTypes != null && i < argTypes.size())
				{
					TypeEnum metricArg = argTypes.get(i);
					paramType = generator.transformType(metricArg, null, param);
				}

				// System.err.println("****** paramType = " + paramType + " value = " +
				// param.getValue());
				generator.addMethodParameter(paramType, param.var.getValue());
			}
		}

		// addLocalVars(transformer, generator);

		Collection<AbstractStatement> newStmts = transformer.transformStatement(generator, block);
		if (newStmts != null)
		{
			for (AbstractStatement newStmt : newStmts)
			{
				generator.addStatement(newStmt, block);
			}
		}

		generator.doneMethod();
	}

//	private boolean isFuncParam(String name)
//	{
//		if (funcParamDefs != null && funcParamDefs.isPresent())
//		{
//			int numParams = funcParamDefs.getPrimaryCount();
//			for (int i = 0; i < numParams; i++)
//			{
//				Rust_Parameter param = funcParamDefs.getPrimaryElement(i);
//				Rust_Variable_Definition var = param.var;
//				if (var.getValue().equalsIgnoreCase(name))
//				{
//					return true;
//				}
//			}
//		}
//		return false;
//	}

//	// Are there any local variables we need to declare?
//	private void addLocalVars(EagleTransformer transformer,
//			EagleGenerator<AbstractStatement, AbstractExpression, AbstractVariable, AbstractType> generator)
//	{
//		String scopeStr = this._currentLine + "-" + this._endLine;
//		ArrayList<AssignMetrics> asgMetrics = transformer._metrics.findVarsInScope(scopeStr);
//		for (AssignMetrics met : asgMetrics)
//		{
//			TypeEnum typ = met.uniqueType();
//			if (typ != TypeEnum.VOID)
//			{
//				if (!isFuncParam(met._symbolName))
//				{
//					// System.err.println("****** Found var " + met._symbolName);
//					AbstractType absType = generator.transformType(typ, null, this);
//					AbstractStatement dataStmt = generator.newDataDeclaration(StaticEnum.NONE,
//							met._symbolName, null, absType, null, this);
//					generator.addStatement(dataStmt, this);
//				}
//			}
//		}
//	}

	public void newRustFunction(Rust_Type returnType, String methodName)
	{
		this.leftParen = new PunctuationLeftParen();
		this.rightParen = new PunctuationRightParen();

		if (returnType != null)
		{
			this.returns = new Rust_FunctionReturns();
			this.returns.setPresent(true);
			this.returns.returnType = returnType;
		}

		this.block = new Rust_Block_Statement();
		this.block.leftBrace = new PunctuationLeftBrace();
		this.block.statements = new TokenList<Rust_Statement>();
		this.block.rightBrace = new PunctuationRightBrace();

		this.id = new Rust_Function_Definition();
		this.id.setValue(methodName);
	}

	public void addFunctionParameter(Rust_Type type, String name)
	{
		Rust_Parameter param = new Rust_Parameter();
		param.var = new Rust_Variable_Definition();
		param.var.setValue(name);
		param.colon = new PunctuationColon();
		param.type = type;
		
		if (type.getWhich() instanceof Rust_TypePrimitive)
		{
			Rust_TypePrimitive prim = (Rust_TypePrimitive) type.getWhich();
			if (prim.PRIMITIVE.getValue().equals("String"))
			{
				// &str is a read-only pointer, String is mutable
				// Don't ever want to change function parameters
				prim.PRIMITIVE.setValue("&str");
			}
		}
		
		if (funcParamDefs == null)
		{
			this.funcParamDefs = new SeparatedList<Rust_Parameter, PunctuationComma>();
			this.funcParamDefs.setPresent(true);
		}
		
		if (this.funcParamDefs.size() > 0)
		{
			this.funcParamDefs.addSecondaryElement(new PunctuationComma());
		}
		this.funcParamDefs.addPrimaryElement(param);
	}
	
	public void addComment(Rust_Comment comm)
	{
		Rust_Statement stmt = new Rust_Statement();
		stmt.setWhich(comm);
		this.block.statements._elements.add(stmt);
	}
}

File: csrc/com/eagle/programmar/Rust/Rust_Function.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// Purchase a Premium Edition license at:
// https://www.tangiblesoftwaresolutions.com/order/order-java-to-csharp.html
// ====================================================================================================

using System;
using System.Collections.Generic;

// Copyright Eagle Legacy Modernization, LLC, 2010-date
// Original author: Steven A. O'Hara, Jul 2, 2022

namespace com.eagle.programmar.Rust
{

	using EagleInterpreter = com.eagle.interpret.EagleInterpreter;
	using EagleRunnable = com.eagle.interpret.EagleRunnable;
	using ArgumentsMetrics = com.eagle.metrics.ArgumentsMetrics;
	using CallMetrics = com.eagle.metrics.CallMetrics;
	using Rust_TypePrimitive = com.eagle.programmar.Rust.Rust_Type.Rust_TypePrimitive;
	using Rust_Block_Statement = com.eagle.programmar.Rust.Statements.Rust_Block_Statement;
	using Rust_Function_Definition = com.eagle.programmar.Rust.Symbols.Rust_Function_Definition;
	using Rust_Variable_Definition = com.eagle.programmar.Rust.Symbols.Rust_Variable_Definition;
	using Rust_Comment = com.eagle.programmar.Rust.Terminals.Rust_Comment;
	using Rust_Keyword = com.eagle.programmar.Rust.Terminals.Rust_Keyword;
	using Rust_Punctuation = com.eagle.programmar.Rust.Terminals.Rust_Punctuation;
	using EagleScope = com.eagle.scope.EagleScope;
	using EagleScopeInterface = com.eagle.scope.EagleScope.EagleScopeInterface;
	using AbstractFunction = com.eagle.tokens.AbstractFunction;
	using SeparatedList = com.eagle.tokens.SeparatedList;
	using TokenList = com.eagle.tokens.TokenList;
	using TokenSequence = com.eagle.tokens.TokenSequence;
	using AbstractExpression = com.eagle.tokens.interfaces.AbstractExpression;
	using AbstractStatement = com.eagle.tokens.interfaces.AbstractStatement;
	using AbstractType = com.eagle.tokens.interfaces.AbstractType;
	using AbstractVariable = com.eagle.tokens.interfaces.AbstractVariable;
	using PunctuationColon = com.eagle.tokens.punctuation.PunctuationColon;
	using PunctuationComma = com.eagle.tokens.punctuation.PunctuationComma;
	using PunctuationLeftBrace = com.eagle.tokens.punctuation.PunctuationLeftBrace;
	using PunctuationLeftParen = com.eagle.tokens.punctuation.PunctuationLeftParen;
	using PunctuationRightBrace = com.eagle.tokens.punctuation.PunctuationRightBrace;
	using PunctuationRightParen = com.eagle.tokens.punctuation.PunctuationRightParen;
	using EagleGenerator = com.eagle.transform.EagleGenerator;
	using TypeEnum = com.eagle.transform.EagleGenerator.TypeEnum;
	using EagleTransformableFunction = com.eagle.transform.EagleTransformableFunction;
	using EagleTransformer = com.eagle.transform.EagleTransformer;

	public class Rust_Function : TokenSequence, EagleRunnable, AbstractFunction, EagleScope.EagleScopeInterface, EagleTransformableFunction
	{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(10) @OPT Rust_Keyword PUB = new com.eagle.programmar.Rust.Terminals.Rust_Keyword("pub");
		public  OPT;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(20) @DOC("items/functions.html") com.eagle.programmar.Rust.Terminals.Rust_Keyword FN = new com.eagle.programmar.Rust.Terminals.Rust_Keyword("fn");
		public @DOC("items/functions.html") Rust_Keyword FN = new Rust_Keyword("fn");
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(30) com.eagle.programmar.Rust.Symbols.Rust_Function_Definition id;
		public Rust_Function_Definition id;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(40) @NOSPACE PunctuationLeftParen leftParen;
		public @NOSPACE PunctuationLeftParen leftParen;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(50) @OPT @NOSPACE SeparatedList<Rust_Parameter, com.eagle.tokens.punctuation.PunctuationComma> funcParamDefs;
		public @OPT SeparatedList<Rust_Parameter, PunctuationComma> funcParamDefs;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(60) @NOSPACE PunctuationRightParen rightParen;
		public @NOSPACE PunctuationRightParen rightParen;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(70) @OPT Rust_FunctionReturns returns;
		public @OPT Rust_FunctionReturns returns;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(80) com.eagle.programmar.Rust.Statements.Rust_Block_Statement block;
		public Rust_Block_Statement block;

		public static class Rust_FunctionReturns extends TokenSequence
		{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(10) com.eagle.programmar.Rust.Terminals.Rust_Punctuation arrow = new com.eagle.programmar.Rust.Terminals.Rust_Punctuation("->");
			public Rust_Punctuation arrow = new Rust_Punctuation("->");
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(20) Rust_Type returnType;
			public Rust_Type returnType;
		}

		public static class Rust_Parameter extends TokenSequence
		{
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(10) com.eagle.programmar.Rust.Symbols.Rust_Variable_Definition var;
			public Rust_Variable_Definition var;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(20) com.eagle.tokens.punctuation.PunctuationColon colon;
			public PunctuationColon colon;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @S(30) Rust_Type type;
			public Rust_Type type;
		}

// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: private @SKIP EagleScope _scope = new com.eagle.scope.EagleScope(this, Rust_Syntax.IS_CASE_SENSITIVE);
		private EagleScope _scope = new EagleScope(this, Rust_Syntax.IS_CASE_SENSITIVE);
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @SKIP CallMetrics _callMetrics = null;
		public CallMetrics _callMetrics = null;
// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: public @SKIP ArgumentsMetrics _argumentsMetrics = null;
		public ArgumentsMetrics _argumentsMetrics = null;

		public EagleScope Scope
		{
			return _scope;
		}

		public void interpret(EagleInterpreter interpreter)
		{
			if (_callMetrics == null)
			{
				_callMetrics = new CallMetrics(interpreter._metrics, id.getValue(), id);
			}
			if (_argumentsMetrics == null)
			{
				_argumentsMetrics = new ArgumentsMetrics(interpreter._metrics, id.getValue(), id);
			}

			// Don't do anything here.
			// We searched for all the functions in a preliminary pass
			// And we only evaluate when it is called

			// Except the function called 'main'
			if (id.getValue().Equals("main"))
			{
				interpreter.callingFunction("main", this);
				interpreter.tryToInterpret(block);
				interpreter.completedFunction("main", this);
			}
		}

		public void transformFunction(EagleTransformer transformer, EagleGenerator<AbstractStatement, AbstractExpression, AbstractVariable, AbstractType> generator)
		{
			AbstractType newReturnType = null;
			if (returns != null && returns.isPresent())
			{
				if (returns.returnType.getWhich() is Rust_TypePrimitive)
				{
					Rust_TypePrimitive primitive = (Rust_TypePrimitive) returns.returnType.getWhich();
					newReturnType = Rust_Type.findType(generator, primitive.PRIMITIVE.ToString());
				}
			}

			if (newReturnType == null)
			{
				EagleGenerator.TypeEnum metricRetType = transformer.findReturnMetric(id);
				newReturnType = generator.transformType(metricRetType, null, id);
			}

			string fnName = id.getValue();
			bool isMain = false;
			if (fnName.Equals("main"))
			{
				fnName = generator.mainName(); // Change from 'main' to 'Main' for C#
				isMain = true;
			}

			generator.addMethod(newReturnType, fnName, this);
			generator.setMethodName(fnName);
			if (VERBOSE)
			{
				Console.WriteLine("** Found Rust function " + fnName);
			}

			if (isMain)
			{
				// Have to wait until addMethod is called
				generator.addMainArgs(); // For java and C# but not for Python or Rust
			}

			// Search metrics for arg types -- might not be any
			List<EagleGenerator.TypeEnum> argTypes = transformer.findArgumentsMetric(id);

			if (funcParamDefs != null && funcParamDefs.isPresent())
			{
				for (int i = 0; i < funcParamDefs.getPrimaryCount(); i++)
				{
					Rust_Parameter param = funcParamDefs.getPrimaryElement(i);
					AbstractType paramType = null;

					if (argTypes != null && i < argTypes.Count)
					{
						EagleGenerator.TypeEnum metricArg = argTypes[i];
						paramType = generator.transformType(metricArg, null, param);
					}

					// System.err.println("****** paramType = " + paramType + " value = " +
					// param.getValue());
					generator.addMethodParameter(paramType, param.var.getValue());
				}
			}

			///////// addLocalVars(transformer, generator);

			ICollection<AbstractStatement> newStmts = transformer.transformStatement(generator, block);
			if (newStmts != null)
			{
				foreach (AbstractStatement newStmt in newStmts)
				{
					generator.addStatement(newStmt, block);
				}
			}

			generator.doneMethod();
		}

		public void newRustFunction(Rust_Type returnType, string methodName)
		{
			this.leftParen = new PunctuationLeftParen();
			this.rightParen = new PunctuationRightParen();

			if (returnType != null)
			{
				this.returns = new Rust_FunctionReturns();
				this.returns.setPresent(true);
				this.returns.returnType = returnType;
			}

			this.block = new Rust_Block_Statement();
			this.block.leftBrace = new PunctuationLeftBrace();
			this.block.statements = new TokenList<Rust_Statement>();
			this.block.rightBrace = new PunctuationRightBrace();

			this.id = new Rust_Function_Definition();
			this.id.setValue(methodName);
		}

		public void addFunctionParameter(AbstractType type, string name)
		{
			Rust_Parameter param = new Rust_Parameter();
			param.var = new Rust_Variable_Definition();
			param.var.setValue(name);
			param.type = (Rust_Type) type;

			if (funcParamDefs == null)
			{
				this.funcParamDefs = new SeparatedList<Rust_Parameter, PunctuationComma>();
				this.funcParamDefs.setPresent(true);
			}

			if (this.funcParamDefs.size() > 0)
			{
				this.funcParamDefs.addSecondaryElement(new PunctuationComma());
			}
			this.funcParamDefs.addPrimaryElement(param);
		}

		public void addComment(Rust_Comment comm)
		{
			Rust_Statement stmt = new Rust_Statement();
			stmt.setWhich(comm);
			this.block.statements._elements.add(stmt);
		}
	}

}
